include/drivers/host_cmd_periph.h

```diff
@@ -0,0 +1,165 @@
+	size_t *len;
+	/** Device will take when it needs to write to @a buf and @a size. */
+	struct k_sem *device_owns;
+	/** Handler will take so it can read @a buf and @a size */
+	struct k_sem *handler_owns;
+};
+
+/**
+ * @brief Context for host command peripheral and framework to pass tx data
+ */
+struct host_cmd_periph_tx_buf {
+	/** Data to write to the host */
+	void *buf;
+	/** Number of bytes to write from @a buf */
+	size_t len;
+};
+
+typedef int (*host_cmd_periph_api_init)(
+	struct device *dev, struct host_cmd_periph_rx_context *rx_ctx);
+
+typedef int (*host_cmd_periph_api_send)(
+	struct device *dev, const struct host_cmd_periph_tx_buf *tx_buf);
+
+__subsystem struct host_cmd_periph_api {
+	host_cmd_periph_api_init init;
+	host_cmd_periph_api_send send;
+};
+
+/**
+ * @brief Initialize a host command device
+ *
+ * This routine initializes a host command device, prior to its first use. The
+ * receive context object are an output of this function and are valid
+ * for the lifetime of this device. The RX context is used by the client to
+ * receive data from the host.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param rx_ctx [out] The receiving context object that are valid for the
+ *               lifetime of the device. These objects are used to receive data
+ *               from the driver when the host send data.
+ *
+ * @retval 0 if successful
+ */
+__syscall int host_cmd_periph_init(struct device *dev,
+				   struct host_cmd_periph_rx_context *rx_ctx);
+
+static inline int
+z_impl_host_cmd_periph_init(struct device *dev,
+			    struct host_cmd_periph_rx_context *rx_ctx)
+{
+	const struct host_cmd_periph_api *api =
+		(const struct host_cmd_periph_api *)dev->api;
+
+	return api->init(dev, rx_ctx);
+}
+
+#ifdef CONFIG_USERSPACE
+static inline void
+z_vrfy_host_cmd_periph_init(struct device *dev,
+			    struct host_cmd_periph_rx_context *rx_ctx)
+{
+	struct host_cmd_periph_rx_context local_rx_ctx;
+
+	z_impl_host_cmd_periph_init(dev, &local_rx_ctx);
+
+	Z_OOPS(z_user_to_copy(&local_rx_ctx, rx_ctx, sizeof(*rx_ctx)));
+}
+#include <syscalls/host_cmd_periph_init_mrsh.c>
+#endif
+
+
+/**
+ * @brief Sends the specified data to the host
+ *
+ * Sends the data specified in @a tx_buf to the host over the host communication
+ * bus.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param tx_buf The data to transmit to the host.
+ *
+ * @retval 0 if successful
+ */
+__syscall int host_cmd_periph_send(struct device *dev,
+				   const struct host_cmd_periph_tx_buf *tx_buf);
+
+static inline int
+z_impl_host_cmd_periph_send(struct device *dev,
+			    const struct host_cmd_periph_tx_buf *tx_buf)
+{
+	const struct host_cmd_periph_api *api =
+		(const struct host_cmd_periph_api *)dev->api;
+
+	return api->send(dev, tx_buf);
+}
+
+#ifdef CONFIG_USERSPACE
+static inline void
+z_vrfy_host_cmd_periph_send(struct device *dev,
```

Comment:
please put the verification functions in a C file and not a header, they must be a singleton. they are declared inline because they get folded into the z_mrsh_ generated function

Reply:
Ah, I see that now. I must have been using timer.c (or something similar) as my example and saw the z_impl and z_vrfy together (because it was the c file), and I didn't realize I was looking at c file instead of h file snippet for my example. Thanks!